python/tvm/relay/frontend/tensorflow.py

```diff
@@ -1166,6 +1166,125 @@ def _impl(inputs, attr, params, mod):
     return _impl
 
 
+def _math_segment_sum():
+    def _impl(inputs, attr, params, mod):
+        assert len(inputs) == 2, "There should be 2 input tensors"
+        return get_relay_op("segment_sum")(inputs[0], inputs[1])
+
+    return _impl
+
+
+def _sparse_segment_sum():
+    def _impl(inputs, attr, params, mod):
+        assert len(inputs) == 3, "There should be 3 input tensors"
+        data = _op.take(inputs[0], inputs[1], axis=0)
+        return _op.segment_sum(data, inputs[2])
+
```

Comment:
This is ok for now, but we definitely want a fused implementation here, just like TF/PT/C2 does. I don't expect this would work for a huge embedding table people want to use in practice.

Reply:
I agree. When you say a "fused implementation" , do you mean that all of it happens in a single ir ?

Reply:
Do you have any examples of what a "fused implementation" is ? Does this mean that in a fused implementation,  the frontend will always just be a one liner ?

Reply:
In this case, I understand we must do the take and the addition from segment sum simultaneously for performance. So a fused implementation in that case would be a new op ?

Reply:
By "fused" I meant we shouldn't materialize the result of `take`, which can be huge. In a fused implementation, we need to look up indices and accumulate the sum on the fly. This is why PT has `EmbeddingBag` op, see their doc https://pytorch.org/docs/stable/generated/torch.nn.EmbeddingBag.html.

Yes, a complicated op like this will not likely be feasible if we rely only on Relay-level op fusion. We need a dedicated `sparse_segment_sum` TOPI/Relay op.

Reply:
I'm not aware of existing ops that I can count as "custom fused op". An efficient implementation of `sparse_segment_sum` is a new challenge for us I think. You can take a look at caffe2 repository for example, they have a LOT of code to better optimize this op. They even have a codegen tool for this op, see https://github.com/pytorch/pytorch/blob/master/caffe2/perfkernels/hp_emblookup_codegen.py and the generated code https://github.com/pytorch/pytorch/blob/master/caffe2/perfkernels/embedding_lookup_fused_8bit_rowwise_avx2.cc. This is the level of seriousness they have for this op.

Reply:
I don't know why it took me so long to realize, but this is just a special case of scatter_nd! You should be able to rewrite all of this in terms of scatter_nd.

Reply:
I am using scatter_add in the relay op but how would you combine take with it to generate one op. I am still trying to find out.

Reply:
AFAIK, Scatter_nd and scatter_add are the same op, just with zero initialization as the difference.

Reply:
I think he meant that `scatter_nd` exactly realized fused `take` and `segment_sum` above. I haven't put deep thought into this but it made sense to me. But I remember parallelizing `scatter_nd` looked harder than `scatter_add`.

Reply:
Yes, I am having a bit of a mind block understanding how take and segment_sum is essentially scatter_nd, do anyone of you mind writing small pseudocode ?

Reply:
FWIW I did a few variants of `torch.nn.EmbeddingBag`, `c2::sparse_length_sum`, etc in TVM IR in https://github.com/ajtulloch/tvm/blob/4b98beb75ca1505ec81ddca358ad61282ab6a05b/topi/python/topi/x86/sparse.py#L162-L257, https://github.com/ajtulloch/tvm/blob/sparse-ops/topi/python/topi/sparse/sparse_lengths_sum.py#L45-L98, https://github.com/ajtulloch/sparse-ads-baselines/blob/a495ea076882615d454d27a1a5b191ec675d3acc/lxu_cache_cpu_funcs.py#L8-L149, etc if that's of interest.

Reply:
Thinking about this more, I believe the take is necessary if we are using scatter_nd. We could make a more generic version of scatter_nd and gather_nd that has indices in both the input and output buffers. That would cover this case.

Reply:
ok I'll merge this as it is then.